### PR TITLE
SCD221 add optional UB calculation in SCDCalibratePanels2 (master)

### LINKED ISSUE #30712: SCD221 add optional UB calculation in SCDCalibratePanels2
The current SCDCalibratePanels2 does not utilize the input lattice constants to update the UB matrix attached to the given peakworkspace.
For more details, please refer to the Gitlab defects [SCD221](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/221).

### PR DESCRIPTION
**Pull request by KedoKudo**
**Description of work.**

Current `SCDCalibratePanels` (version 2) does not utilize the input lattice constants to update the UB matrix attached to the input peak workspace.
This PR addresses this issue and make the update UB optional in case the user would like to use a fixed UB matrix for the calibration process.
Here is a screenshot of the new interface
 
image

> - Related Gitlab task [SCD214](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/214)  
> - This is #30711 into `master`

**To test:**

Use the following code in MantidWorkbench to test `SCDCalibratePanels` (v2)
```python
# necessary import
import numpy as np
from collections import namedtuple
from mantid.simpleapi import *


def convert(dictionary):
    return namedtuple('GenericDict', dictionary.keys())(**dictionary)


import os
directory = os.path.dirname(os.path.realpath(__file__))
lc_natrolite = {
    "a": 18.29,  # A
    "b": 18.64,  # A
    "c": 6.56,  # A
    "alpha": 90,  # deg
    "beta": 90,  # deg
    "gamma": 90,  # deg
}
natrolite = convert(lc_natrolite)
pws = LoadIsawPeaks('data/Natrolite_runs_133752_133812.peaks')

SCDCalibratePanels(
    PeakWorkspace='pws',
    a=natrolite.a,
    b=natrolite.b,
    c=natrolite.c,
    alpha=natrolite.alpha,
    beta=natrolite.beta,
    gamma=natrolite.gamma,
    CalibrateT0=False,
    CalibrateL1=True,
    CalibrateBanks=True,
    OutputWorkspace="testCaliTable",
    DetCalFilename=directory + "/test.DetCal",
    CSVFilename=directory + "/test.csv",
    XmlFilename=directory + "/test.xml",
    ToleranceOfTranslation=0.0005,
    ToleranceOfReorientation=0.05,
    TranslationSearchRadius=0.05,
    RotationSearchRadius=5,
    SourceShiftSearchRadius=0.1,
    VerboseOutput=True,
)
```

> - testing data: [data.tar.gz](https://github.com/mantidproject/mantid/files/5991908/data.tar.gz)

Fixes #30712. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
Look OK but can you handle better this case:

```python
pws = LoadIsawPeaks('data/Natrolite_runs_133752_133812.peaks')
ClearUB('pws')
SCDCalibratePanels(
    PeakWorkspace='pws',
    CalibrateT0=False,
    CalibrateL1=True,
    CalibrateBanks=True,
    OutputWorkspace="testCaliTable",
    DetCalFilename=directory + "/test.DetCal",
    CSVFilename=directory + "/test.csv",
    XmlFilename=directory + "/test.xml",
    ToleranceOfTranslation=0.0005,
    ToleranceOfReorientation=0.05,
    TranslationSearchRadius=0.05,
    RotationSearchRadius=5,
    SourceShiftSearchRadius=0.1,
    VerboseOutput=True,
)
```

I get 

```
SCDCalibratePanels-[Error] Error in execution of algorithm SCDCalibratePanels:
SCDCalibratePanels-[Error] When setting value of property "alpha": Selected value 8.98847e+307 is > the upper bound (175)
ValueError: When setting value of property "alpha": Selected value 8.98847e+307 is > the upper bound (175)
  File "<string>", line 2, in <module>
  File "/home/rwp/src/mantid/Framework/PythonInterface/mantid/simpleapi.py", line 1036, in __call__
    algm.execute()
```

Maybe validate that the UB exist before trying to run the algorithm when the lattice parameters aren't supplied because it won't be clear to the user what is happening. Say something like either provide all lattice parameters or set the UB on the workspace.

**Comment by KedoKudo** [comment]
> Look OK but can you handle better this case:
> 
> ```python
> pws = LoadIsawPeaks('data/Natrolite_runs_133752_133812.peaks')
> ClearUB('pws')
> SCDCalibratePanels(
>     PeakWorkspace='pws',
>     CalibrateT0=False,
>     CalibrateL1=True,
>     CalibrateBanks=True,
>     OutputWorkspace="testCaliTable",
>     DetCalFilename=directory + "/test.DetCal",
>     CSVFilename=directory + "/test.csv",
>     XmlFilename=directory + "/test.xml",
>     ToleranceOfTranslation=0.0005,
>     ToleranceOfReorientation=0.05,
>     TranslationSearchRadius=0.05,
>     RotationSearchRadius=5,
>     SourceShiftSearchRadius=0.1,
>     VerboseOutput=True,
> )
> ```
> 
> I get
> 
> ```
> SCDCalibratePanels-[Error] Error in execution of algorithm SCDCalibratePanels:
> SCDCalibratePanels-[Error] When setting value of property "alpha": Selected value 8.98847e+307 is > the upper bound (175)
> ValueError: When setting value of property "alpha": Selected value 8.98847e+307 is > the upper bound (175)
>   File "<string>", line 2, in <module>
>   File "/home/rwp/src/mantid/Framework/PythonInterface/mantid/simpleapi.py", line 1036, in __call__
>     algm.execute()
> ```
> 
> Maybe validate that the UB exist before trying to run the algorithm when the lattice parameters aren't supplied because it won't be clear to the user what is happening.

That is a good catch.
I will add a validation in the `validatInputs()` to make sure this edge case won't pass the check.

File: Framework/Crystal/src/SCDCalibratePanels2ObjFunc.cpp [hunk 1]
```diff
@@ -143,8 +143,8 @@ void SCDCalibratePanels2ObjFunc::function1D(double *out, const double *xValues,
     V3D hkl =
         V3D(boost::math::iround(pk.getH()), boost::math::iround(pk.getK()),
             boost::math::iround(pk.getL()));
-    if (hkl == UNSET_HKL)
-      throw std::runtime_error("Found unindexed peak in input workspace!");
+    // if (hkl == UNSET_HKL)
+    //   throw std::runtime_error("Found unindexed peak in input workspace!");
```
Comment (rosswhitfield) [inline Framework/Crystal/src/SCDCalibratePanels2ObjFunc.cpp:147]: Is this meant to be commented out?

Comment (KedoKudo) [reply Framework/Crystal/src/SCDCalibratePanels2ObjFunc.cpp:147]: Apparently there are un-indexed peaks in the peak workspace after passing it through IndexPeaks, so I have to comment them out.

Comment (rosswhitfield) [reply Framework/Crystal/src/SCDCalibratePanels2ObjFunc.cpp:147]: So should this be removed? Or will this be addressed in a later PR?

Comment (KedoKudo) [reply Framework/Crystal/src/SCDCalibratePanels2ObjFunc.cpp:147]: This should be addressed when we get to changing the objective function.  I would like to keep it in as comment for now as a reminder.

**Review by rosswhitfield — APPROVED** [review]

File: Framework/Crystal/test/SCDCalibratePanels2Test.h [hunk 2]
```diff
@@ -497,8 +497,8 @@ class SCDCalibratePanels2Test : public CxxTest::TestSuite {
     sub_alg->initialize();
     sub_alg->setLogging(LOGCHILDALG);
     sub_alg->setProperty("Workspace", wsname);
-    sub_alg->setProperty("u", "1,0,0");
-    sub_alg->setProperty("v", "0,1,0");
+    sub_alg->setProperty("u", "0.5,0.8660254037844387,0");
```
Comment (peterfpeterson) [inline Framework/Crystal/test/SCDCalibratePanels2Test.h:500]: Much better u,v vectors to test with. Thanks!

**Pull request merged**